lib/hls/hls_parser.js

```diff
@@ -1966,6 +2034,69 @@ shaka.hls.HlsParser = class {
       references.push(reference);
     }
 
+    // If some segments have sync times, but not all, extrapolate the sync
+    // times of the ones with none.
+    const someSyncTime = references.some((ref) => ref.syncTime != null);
+    if (someSyncTime) {
+      for (let i = 0; i < references.length; i++) {
+        const reference = references[i];
+        if (reference.syncTime != null) {
+          // No need to extrapolate.
+          continue;
+        }
+        // Find the nearest segment with syncTime, in either direction.
+        // This looks forward and backward simultaneously, keeping track of what
+        // to offset the syncTime it finds by as it goes.
+        let forwardAdd = 0;
+        let forwardI = i;
+        /** @return {?number} */
+        const lookForward = () => {
+          const other = references[forwardI];
+          if (other) {
+            if (other.syncTime != null) {
+              return other.syncTime + forwardAdd;
+            }
+            forwardAdd -= other.endTime - other.startTime;
```

Comment:
The asymmetry between forward and backward is necessary, but a reader might think you had done that by mistake.  I suggest you add a comment here such as "looking forward, sum all durations, including the original reference at `i`, excluding the `other` with a `syncTime`."

Similarly, below I would add a comment like "looking backward, sum all durations, including the `other` with a `syncTime`, but excluding the original reference at `i`."

Reply:
Added comments.